File: src/main/java/br/unifor/costify/application/usecase/RegisterIngredientUseCase.java
package br.unifor.costify.application.usecase;

import br.unifor.costify.application.contracts.IngredientRepository;
import br.unifor.costify.application.dto.command.RegisterIngredientCommand;
import br.unifor.costify.application.dto.entity.IngredientDto;
import br.unifor.costify.application.errors.IngredientAlreadyExistsException;
import br.unifor.costify.application.factory.IngredientFactory;
import br.unifor.costify.application.validation.ValidationService;
import br.unifor.costify.domain.entity.Ingredient;
import org.springframework.stereotype.Service;

@Service
public class RegisterIngredientUseCase {
  private final IngredientRepository ingredientRepository;
  private final IngredientFactory ingredientFactory;
  private final ValidationService validationService;

  public RegisterIngredientUseCase(
      IngredientRepository ingredientRepository, 
      IngredientFactory ingredientFactory,
      ValidationService validationService) {
    this.ingredientRepository = ingredientRepository;
    this.ingredientFactory = ingredientFactory;
    this.validationService = validationService;
  }

  public IngredientDto execute(RegisterIngredientCommand command) {
    validationService.validateIngredientData(
        command.name(), 
        command.packageQuantity(), 
        command.packagePrice(), 
        command.packageUnit()
    );
    
    validateIngredientDoesNotExist(command.name());

    Ingredient ingredient =
        ingredientFactory.create(
            command.name(),
            command.packageQuantity(),
            command.packagePrice(),
            command.packageUnit());

    Ingredient savedIngredient = ingredientRepository.save(ingredient);

    return IngredientDto.from(savedIngredient);
  }

  private void validateIngredientDoesNotExist(String name) {
    if (ingredientRepository.existsByName(name)) {
      throw new IngredientAlreadyExistsException(
          "Ingredient with name '" + name + "' already exists");
    }
  }
}

File: internal/application/usecase/register_ingredient_usecase.go
package usecase

import (
	"github.com/vini/costify-go/internal/application/contract"
	"github.com/vini/costify-go/internal/application/dto/command"
	"github.com/vini/costify-go/internal/application/dto/entity"
	"github.com/vini/costify-go/internal/application/errors"
	"github.com/vini/costify-go/internal/application/factory"
)

// RegisterIngredientUseCase handles the registration of new ingredients
type RegisterIngredientUseCase struct {
	ingredientRepository contract.IngredientRepository
	ingredientFactory    *factory.IngredientFactory
}

// NewRegisterIngredientUseCase creates a new RegisterIngredientUseCase
func NewRegisterIngredientUseCase(
	ingredientRepository contract.IngredientRepository,
	ingredientFactory *factory.IngredientFactory,
) *RegisterIngredientUseCase {
	return &RegisterIngredientUseCase{
		ingredientRepository: ingredientRepository,
		ingredientFactory:    ingredientFactory,
	}
}

// Execute executes the register ingredient use case
func (uc *RegisterIngredientUseCase) Execute(cmd command.RegisterIngredientCommand) (entity.IngredientDto, error) {
	// Validate that ingredient doesn't already exist
	exists, err := uc.ingredientRepository.ExistsByName(cmd.Name)
	if err != nil {
		return entity.IngredientDto{}, err
	}
	if exists {
		return entity.IngredientDto{}, errors.NewIngredientAlreadyExistsError(
			"Ingredient with name '" + cmd.Name + "' already exists")
	}

	// Create ingredient using factory
	ingredient, err := uc.ingredientFactory.Create(
		cmd.Name,
		cmd.PackageQuantity,
		cmd.PackagePrice,
		cmd.PackageUnit,
	)
	if err != nil {
		return entity.IngredientDto{}, err
	}

	// Save ingredient
	savedIngredient, err := uc.ingredientRepository.Save(ingredient)
	if err != nil {
		return entity.IngredientDto{}, err
	}

	return entity.FromIngredient(savedIngredient), nil
}
